src/compiler/builder.ts

```diff
@@ -1099,6 +1199,56 @@ namespace ts {
             );
         }
 
+        function getWriteFileCallback(writeFile: WriteFileCallback | undefined, customTransformers: CustomTransformers | undefined): WriteFileCallback {
+            return (fileName, text, writeByteOrderMark, onError, sourceFiles, data) => {
+                if (isDeclarationFileName(fileName)) {
+                    if (!outFile(state.compilerOptions)) {
+                        Debug.assert(sourceFiles?.length === 1);
+                        let newSignature;
+                        if (!customTransformers) {
+                            const file = sourceFiles[0];
+                            const info = state.fileInfos.get(file.resolvedPath)!;
+                            const signature = state.currentAffectedFilesSignatures?.get(file.resolvedPath) || info.signature;
+                            if (signature === file.version) {
+                                newSignature = computeSignature(text, data, computeHash);
+                                if (newSignature !== file.version) { // Update it
+                                    if (host.storeFilesChangingSignatureDuringEmit) (state.filesChangingSignature ||= new Set()).add(file.resolvedPath);
+                                    if (state.exportedModulesMap) BuilderState.updateExportedModules(file, file.exportedModulesFromDeclarationEmit, state.currentAffectedFilesExportedModulesMap ||= BuilderState.createManyToManyPathMap());
+                                    if (state.affectedFiles && state.affectedFilesIndex! < state.affectedFiles.length) {
+                                        state.currentAffectedFilesSignatures!.set(file.resolvedPath, newSignature);
+                                    }
+                                    else {
+                                        info.signature = newSignature;
+                                        if (state.exportedModulesMap) BuilderState.updateExportedFilesMapFromCache(state, state.currentAffectedFilesExportedModulesMap);
+                                    }
+                                }
+                            }
+                        }
+
+                        if (state.compilerOptions.composite) {

```

Comment:
This part is where we store the d.ts emit hash so that we can compare if d.ts has changed. We do this only for composite projects because only these can be referenced in other projects and need the d.ts change time to determine if update to the projects referencing them is needed or just timestamp updates are needed